File: src/data_structure/linked_list/getnode.cpp
/*
  Get Nth element from the end in a linked list of integers
  Number of elements in the list will always be greater than N.
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/
int GetNode(Node *head,int positionFromTail)
{
    Node *p, *q, *r;
    if (head == NULL)
    {
        return 0;
    }
    if (head->next == NULL)
    {
        return head->data;
    }
    p = head;
    q = head->next;
    p->next = NULL;
    do
    {
        r = q->next;
        q->next = p;
        p = q;
        q = r;
    } while(r != NULL);

    Node *reverse_head = p;
    int i = 0;
    while (reverse_head != NULL && i++ < positionFromTail)
    {
        reverse_head = reverse_head->next;
    }
    return reverse_head->data;
}

File: src/data_structure/linked_list/getnode.java
/*
  Insert Node at the end of a linked list 
  head pointer input could be NULL as well for empty list
  Node is defined as 
  class Node {
     int data;
     Node next;
  }
*/
    
int GetNode(Node head,int n) {

    if (head == null) {
        return 0;
    }
    if (head.next == null) {
        return head.data;
    }
    Stack<Node> stack = new Stack<Node>();
    while (head != null) {
        stack.push(head);
        head = head.next;
    }
    
    for (int i = 0; i<n; i++) {
        stack.pop();
    }
    return stack.pop().data;
}

